### PR TITLE
The nuclear option

### PR DESCRIPTION
**Pull request by morrowwolf**
# About the pull request

This PR adds a nuke as something buyable at the tier 4 level of intel. It takes 35 points in total to reach it (costing 20 points by itself). You cannot buy it before 120 minutes.

To use the nuke you must first decrypt the authorization codes on the surface. This requires holding both comms towers on the surface for 10 minutes.

After it's decrypted you are free to arm the nuke, without requiring the comm towers, and it will detonate in 60 seconds.

# Explain why it's good for the game

Coming soonTM

# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog
:cl: Morrow
add: Re-added the nuke that you can buy via intel at tier 4. Only purchasable past 120 minutes. Requires holding both groundside comms towers for authorization decryption before being able to be armed.
/:cl:

### CODE REVIEW

File: code/game/machinery/nuclearbomb.dm [hunk 1]
```diff
@@ -55,39 +55,41 @@ var/bomb_set = FALSE
...
-	else
+	if(!timing)
+		update_minimap_icon()
 		stop_processing()
+		return
+
+	bomb_set = TRUE //So long as there is one nuke timing, it means one nuke is armed.
+	timeleft = explosion_time - world.time
+	if(world.time >= explosion_time)
+		explode()
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:66]:
```suggestion
		explode()
		return
```

File: code/game/machinery/nuclearbomb.dm [hunk 2]
```diff
@@ -55,39 +55,41 @@ var/bomb_set = FALSE
...
+	//3 warnings: 1. Halfway through, 2. 1 minute left, 3. 10 seconds left.
+	//this structure allows varedits to var/timeleft without losing or spamming warnings.
+	else if(timer_announcements_flags)
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:69]:
```suggestion
	if(!timer_announcements_flags)
		return
```

File: code/game/machinery/nuclearbomb.dm [hunk 3]
```diff
@@ -55,39 +55,41 @@ var/bomb_set = FALSE
...
+		if(timer_announcements_flags & NUKE_SHOW_TIMER_MINUTE)
+			if(timeleft <= 600 && timeleft >= 570)
+				announce_to_players(NUKE_SHOW_TIMER_MINUTE)
+				timer_announcements_flags = NUKE_SHOW_TIMER_TEN_SEC
+				return
+		if(timer_announcements_flags & NUKE_SHOW_TIMER_TEN_SEC)
+			if(timeleft <= 100 && timeleft >= 70)
+				announce_to_players(NUKE_SHOW_TIMER_TEN_SEC)
+				timer_announcements_flags = 0
+				return
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:84]:
```suggestion
	if(timer_announcements_flags & NUKE_SHOW_TIMER_HALF)
		if(timeleft <= initial(timeleft) / 2 && timeleft >= initial(timeleft) / 2 - 30)
			announce_to_players(NUKE_SHOW_TIMER_HALF)
			timer_announcements_flags &= ~NUKE_SHOW_TIMER_HALF
			return
	if(timer_announcements_flags & NUKE_SHOW_TIMER_MINUTE)
		if(timeleft <= 600 && timeleft >= 570)
			announce_to_players(NUKE_SHOW_TIMER_MINUTE)
			timer_announcements_flags = NUKE_SHOW_TIMER_TEN_SEC
			return
	if(timer_announcements_flags & NUKE_SHOW_TIMER_TEN_SEC)
		if(timeleft <= 100 && timeleft >= 70)
			announce_to_players(NUKE_SHOW_TIMER_TEN_SEC)
			timer_announcements_flags = 0
			return
```

File: code/game/machinery/nuclearbomb.dm [hunk 4]
```diff
@@ -422,3 +407,188 @@ var/bomb_set = FALSE
 	bomb_set = FALSE
 	SSminimaps.remove_marker(src)
 	return ..()
+
+/obj/structure/machinery/nuclearbomb/tech
+
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:412]:
```suggestion
```

**Comment by morrowwolf** [comment]
Some notes for reviewers:
I need an intel icon (sprites)
I need to announce when decryption is done
I need to make the timer the same as the other timer for both arming and destroying
Check if there's a double announce at the beginning of arming
Make window more in like height-wise

File: code/game/machinery/nuclearbomb.dm [hunk 5]
```diff
@@ -422,3 +407,188 @@ var/bomb_set = FALSE
...
+	for(var/obj/structure/machinery/telecomms/relay/preset/tower/mapcomms/telecomm_unit in linked_decryption_towers)
+		RegisterSignal(telecomm_unit, COMSIG_COMM_RELAY_SHUT_DOWN, PROC_REF(connected_comm_shutdown))
+
+/obj/structure/machinery/nuclearbomb/tech/ui_data(mob/user)
+	var/list/data = ..()
+
+	data["decrypting"] = decrypting
+	data["decryption_time"] = decryption_time / (1 SECONDS)
+
+	return data
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:445]:
```suggestion
	. = ..()

	.["decrypting"] = decrypting
	.["decryption_time"] = decryption_time / (1 SECONDS)
```

File: code/game/machinery/nuclearbomb.dm [hunk 6]
```diff
@@ -422,3 +407,188 @@ var/bomb_set = FALSE
...
+				return
+
+			if(!anchored)
+				to_chat(ui.user, SPAN_INFO("Engage anchors first!"))
+				return
+
+			var/area/A = get_area(src)
+			if(!A.can_build_special)
+				to_chat(ui.user, SPAN_INFO("You cannot deploy [src] here!"))
+				return
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:467]:
```suggestion
			var/area/current_area = get_area(src)
			if(!current_area.can_build_special)
				to_chat(ui.user, SPAN_INFO("You cannot deploy [src] here!"))
				return
```

File: code/game/machinery/nuclearbomb.dm [hunk 7]
```diff
@@ -422,3 +407,188 @@ var/bomb_set = FALSE
...
+	decryption_time = decryption_end_time - world.time
+
+	if(world.time > decryption_end_time)
+		decrypting = FALSE
+		decryption_time = 0
+		//add announcemnt for end of decryption
+		stop_processing()
+		return
+
+	else if(timer_announcements_flags)
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:511]:
```suggestion
	if(!timer_announcements_flags)
```

File: code/game/machinery/nuclearbomb.dm [hunk 8]
```diff
@@ -422,3 +407,188 @@ var/bomb_set = FALSE
...
+		if(timer_announcements_flags & NUKE_DECRYPT_SHOW_TIMER_HALF)
+			if(decryption_time <= initial(decryption_time) / 2 && decryption_time >= initial(decryption_time) / 2 - 30)
+				announce_to_players(NUKE_DECRYPT_SHOW_TIMER_HALF)
+				timer_announcements_flags &= ~NUKE_DECRYPT_SHOW_TIMER_HALF
+				return
+		if(timer_announcements_flags & NUKE_DECRYPT_SHOW_TIMER_MINUTE)
+			if(decryption_time <= 600 && decryption_time >= 570)
+				announce_to_players(NUKE_DECRYPT_SHOW_TIMER_MINUTE)
+				timer_announcements_flags &= ~NUKE_DECRYPT_SHOW_TIMER_MINUTE
+				return
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:521]:
```suggestion
	if(timer_announcements_flags & NUKE_DECRYPT_SHOW_TIMER_HALF)
		if(decryption_time <= initial(decryption_time) / 2 && decryption_time >= initial(decryption_time) / 2 - 30)
			announce_to_players(NUKE_DECRYPT_SHOW_TIMER_HALF)
			timer_announcements_flags &= ~NUKE_DECRYPT_SHOW_TIMER_HALF
			return
	if(timer_announcements_flags & NUKE_DECRYPT_SHOW_TIMER_MINUTE)
		if(decryption_time <= 600 && decryption_time >= 570)
			announce_to_players(NUKE_DECRYPT_SHOW_TIMER_MINUTE)
			timer_announcements_flags &= ~NUKE_DECRYPT_SHOW_TIMER_MINUTE
			return
```

File: code/game/machinery/nuclearbomb.dm [hunk 9]
```diff
@@ -422,3 +407,188 @@ var/bomb_set = FALSE
...
+	if(timer_warning)
+		announcement_helper("DECRYPTION IN [round(decryption_time/10)] SECONDS.", "[MAIN_AI_SYSTEM] Nuclear Tracker", humans_USCM, 'sound/misc/notice1.ogg')
+		announcement_helper("DECRYPTION IN [round(decryption_time/10)] SECONDS.", "HQ Intel Division", humans_other, 'sound/misc/notice1.ogg')
+
+		//preds part
+		var/time_left = duration2text_sec(round(rand(decryption_time - decryption_time / 10, decryption_time + decryption_time / 10)))
+		yautja_announcement(SPAN_YAUTJABOLDBIG("WARNING!\n\nYou have approximately [time_left] seconds to abandon the hunting grounds before human Purification Device is able to be activated."))
+
+		//xenos part
+		var/warning
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:548]:
```suggestion
		var/warning = "Hive killer is almost prepared to be activated!"
```

File: code/game/machinery/nuclearbomb.dm [hunk 10]
```diff
@@ -422,3 +407,188 @@ var/bomb_set = FALSE
...
+		if(timer_warning & NUKE_DECRYPT_SHOW_TIMER_HALF)
+			warning = "Hive killer is halfway through its initial phase!"
+		else
+			warning = "Hive killer is almost prepared to be activated!"
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:552]:
```suggestion
```

File: code/game/machinery/nuclearbomb.dm [hunk 11]
```diff
@@ -422,3 +407,188 @@ var/bomb_set = FALSE
...
+	if(decrypting)
+		announcement_helper("ALERT.\n\nNUCLEAR EXPLOSIVE ORDNANCE DECRYPTION STARTED.\n\nDECRYPTION IN [round(decryption_time/10)] SECONDS.", "[MAIN_AI_SYSTEM] Nuclear Tracker", humans_USCM, 'sound/misc/notice1.ogg')
+		announcement_helper("ALERT.\n\nNUCLEAR EXPLOSIVE ORDNANCE DECRYPTION STARTED.\n\nDECRYPTION IN [round(decryption_time/10)] SECONDS.", "HQ Nuclear Tracker", humans_other, 'sound/misc/notice1.ogg')
+		var/time_left = duration2text_sec(round(rand(decryption_time - decryption_time / 10, decryption_time + decryption_time / 10)))
+		yautja_announcement(SPAN_YAUTJABOLDBIG("WARNING!<br>A human Purification Device has been detected. You have approximately [time_left] before it finishes its initial phase."))
+		for(var/hivenumber in GLOB.hive_datum)
+			hive = GLOB.hive_datum[hivenumber]
+			if(!hive.totalXenos.len)
+				continue
+			xeno_announcement(SPAN_XENOANNOUNCE("The tallhosts have started the initial phase of a hive killer at [get_area_name(loc)]! Destroy their communications relays!"), hive.hivenumber, XENO_GENERAL_ANNOUNCE)
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:572]:
```suggestion
			xeno_announcement(SPAN_XENOANNOUNCE("The tallhosts have started the initial phase of a hive killer at [get_area_name(loc)]! Destroy their communications relays!"), hive.hivenumber, XENO_GENERAL_ANNOUNCE)
			return
```

File: code/game/machinery/nuclearbomb.dm [hunk 12]
```diff
@@ -422,3 +407,188 @@ var/bomb_set = FALSE
...
+	else
+		announcement_helper("ALERT.\n\nNUCLEAR EXPLOSIVE DECRYPTION HALTED.", "[MAIN_AI_SYSTEM] Nuclear Tracker", humans_USCM, 'sound/misc/notice1.ogg')
+		announcement_helper("ALERT.\n\nNUCLEAR EXPLOSIVE DECRYPTION HALTED.", "HQ Intel Division", humans_other, 'sound/misc/notice1.ogg')
+		yautja_announcement(SPAN_YAUTJABOLDBIG("WARNING!<br>The human Purification Device's signature has disappeared."))
+		for(var/hivenumber in GLOB.hive_datum)
+			hive = GLOB.hive_datum[hivenumber]
+			if(!hive.totalXenos.len)
+				continue
+			xeno_announcement(SPAN_XENOANNOUNCE("The hive killer's initial phase has been halted! Rejoice!"), hive.hivenumber, XENO_GENERAL_ANNOUNCE)
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:581]:
```suggestion

	announcement_helper("ALERT.\n\nNUCLEAR EXPLOSIVE DECRYPTION HALTED.", "[MAIN_AI_SYSTEM] Nuclear Tracker", humans_USCM, 'sound/misc/notice1.ogg')
	announcement_helper("ALERT.\n\nNUCLEAR EXPLOSIVE DECRYPTION HALTED.", "HQ Intel Division", humans_other, 'sound/misc/notice1.ogg')
	yautja_announcement(SPAN_YAUTJABOLDBIG("WARNING!<br>The human Purification Device's signature has disappeared."))
	for(var/hivenumber in GLOB.hive_datum)
		hive = GLOB.hive_datum[hivenumber]
		if(!length(hive.totalXenos))
			continue
		xeno_announcement(SPAN_XENOANNOUNCE("The hive killer's initial phase has been halted! Rejoice!"), hive.hivenumber, XENO_GENERAL_ANNOUNCE)
```

File: code/modules/cm_tech/techs/marine/tier4/nuke.dm [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+/datum/tech/nuke/on_unlock()
+	. = ..()
+
+	var/datum/supply_order/O = new /datum/supply_order()
+	O.ordernum = supply_controller.ordernum
+	supply_controller.ordernum++
+	O.object = supply_controller.supply_packs["Intel Operational Nuke"]
+	O.orderedby = MAIN_AI_SYSTEM
+
+	supply_controller.shoppinglist += O
```
Comment (harryob) [inline code/modules/cm_tech/techs/marine/tier4/nuke.dm:30]:
```suggestion
	var/datum/supply_order/new_order = new /datum/supply_order()
	new_order.ordernum = supply_controller.ordernum
	supply_controller.ordernum++
	new_order.object = supply_controller.supply_packs["Intel Operational Nuke"]
	new_order.orderedby = MAIN_AI_SYSTEM

	supply_controller.shoppinglist += new_order
```

**Review by harryob — CHANGES_REQUESTED** [review]
initial

**Review by fira — CHANGES_REQUESTED** [review]
SPAN_DANGERs weren't so bad given the magnitude and importance of this thing...

**Comment by morrowwolf** [comment]
I'm not ignoring reviews FWIW just trying to settle if this is something I even want to finalize into a hard feature or not before I sink more time on serious polish. I appreciate you all 🙏

**Comment by frozentsbgg** [comment]
90 Minutes is too short.
Probably needs to be 120.

Aiming for the 2:30-3:00 round stalemate breaker is the best function of the nuke.
Good changes making it buyable though!

**Comment by morrowwolf** [comment]
90 minutes is for testing both feasibility and functionality.

We've already learned of a comms location on soro that absolutely needs moved.

**Comment by frozentsbgg** [comment]
> 

Makes sense :) Appreciate the clarification.
Just wanted to state that I like the feature, all the criticism I've seen so far has been time-related, so that makes me feel better.

**Comment by morrowwolf** [comment]
Basic review done, gonna finish up the above polish stuff when I get da sprite. 🙏

File: code/game/machinery/nuclearbomb.dm [hunk 14]
```diff
@@ -422,3 +408,189 @@ var/bomb_set = FALSE
...
+		return ..()
+
+	decryption_time = decryption_end_time - world.time
+
+	if(world.time > decryption_end_time)
+		decrypting = FALSE
+		decryption_time = 0
+		//add announcemnt for end of decryption
+		stop_processing()
+		return
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:509]: return PROCESS_KILL?

**Review by harryob — CHANGES_REQUESTED** [review]

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/game/machinery/nuclearbomb.dm [hunk 15]
```diff
@@ -422,3 +407,208 @@ var/bomb_set = FALSE
...
+	if(!decrypting)
+		return
+
+	decrypting = FALSE
+	announce_to_players()
+
+/obj/structure/machinery/nuclearbomb/tech/Destroy()
+	UnregisterSignal(SSdcs, COMSIG_GLOB_GROUNDSIDE_TELECOMM_TURNED_OFF)
+
+	. = ..()
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:614]:
```suggestion
	return ..()
```

File: code/game/machinery/nuclearbomb.dm [hunk 16]
```diff
@@ -422,3 +407,208 @@ var/bomb_set = FALSE
...
+	SIGNAL_HANDLER
+
+	if(!decrypting)
+		return
+
+	decrypting = FALSE
+	announce_to_players()
+
+/obj/structure/machinery/nuclearbomb/tech/Destroy()
+	UnregisterSignal(SSdcs, COMSIG_GLOB_GROUNDSIDE_TELECOMM_TURNED_OFF)
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:612]: should be unnecessary to handle signal cleanup in destroy

File: code/game/machinery/nuclearbomb.dm [hunk 17]
```diff
@@ -422,3 +407,208 @@ var/bomb_set = FALSE
...
+/obj/structure/machinery/nuclearbomb/tech/ui_data(mob/user)
+	. = ..()
+
+	.["decrypting"] = decrypting
+	.["decryption_time"] = duration2text_sec(decryption_time)
+
+	if(decryption_time)
+		.["decryption_complete"] = FALSE
+	else
+		.["decryption_complete"] = TRUE
```
Comment (harryob) [inline code/game/machinery/nuclearbomb.dm:446]:
```suggestion
	.["decryption_complete"] = decryption_time ? FALSE : TRUE
```

**Review by harryob — CHANGES_REQUESTED** [review]

**Review by harryob — APPROVED** [review]

**Pull request merged**